src/hotspot/share/utilities/vmError.cpp

```diff
@@ -1629,27 +1628,42 @@ void VMError::report_and_die(int id, const char* message, const char* detail_fmt
         // Watcherthread is about to call os::die. Lets just wait.
         os::infinite_sleep();
       } else {
-        // Crash or assert during error reporting. Lets continue reporting with the next step.
-        stringStream ss(buffer, sizeof(buffer));
+        // A secondary error happened. Print brief information, but take care, since crashing
+        // here would just recurse endlessly.
+        // Any information (signal, context, siginfo etc) printed here should use the function
+        // arguments, not the information stored in *this, since those describe the primary crash.
+        char tmp[256]; // cannot use global scratch buffer
         // Note: this string does get parsed by a number of jtreg tests,
         // see hotspot/jtreg/runtime/ErrorHandling.
-        ss.print("[error occurred during error reporting (%s), id 0x%x",
+        st->print("[error occurred during error reporting (%s), id 0x%x",
                    _current_step_info, id);
-        char signal_name[64];
-        if (os::exception_name(id, signal_name, sizeof(signal_name))) {
-          ss.print(", %s (0x%x) at pc=" PTR_FORMAT, signal_name, id, p2i(pc));
+        if (os::exception_name(id, tmp, sizeof(tmp))) {
+          st->print(", %s (0x%x) at pc=" PTR_FORMAT, tmp, id, p2i(pc));
         } else {
           if (should_report_bug(id)) {
-            ss.print(", Internal Error (%s:%d)",
+            st->print(", Internal Error (%s:%d)",
               filename == NULL ? "??" : filename, lineno);
           } else {
-            ss.print(", Out of Memory Error (%s:%d)",
+            st->print(", Out of Memory Error (%s:%d)",
               filename == NULL ? "??" : filename, lineno);
           }
         }
-        ss.print("]");
-        st->print_raw_cr(buffer);
-        st->cr();
+        st->print_cr("]");
+#ifdef ASSERT
+        if (ErrorLogSecondaryErrorDetails) {
```

Comment:
> But printing callstacks during error reporting is unsafe - if we get a second crash or assert, it will cause infinite recursion and interrupt error reporting.

Should there be logic here that at least guarantee progress?
```c++
if (ErrorLogSecondaryErrorDetails && !_some_bool) {
  _some_bool = true;
  [...]
}
_some_bool = false;
```

Reply:
It would work horizontally too, though, since it would disable call stack printing for the next secondary error.

I still think yours is a good idea. We can extend it to using a counter and only allow a small number (2-5) of secondary errors to get printed out with their call stacks. Because if we have more than, say, 5 recursive errors, chances are they all trip over the same thing anyway and more call stacks won't tell you anything new.

And then, if we limit the number of these things, maybe we can afford to leave this feature on by default in debug builds. That addresses the concern  @dholmes-ora voiced, about having to switch on the feature first.

Reply:
>  since it would disable call stack printing for the next secondary error.

 Would it? The example above would only disable reentering the if scope if we have a crash in `[...]`. If it is fine the flag is reset, and if we crash in `[...]` it will not enter the if scope and reset the flag. When it crash in a new secondary error it will once again enter the if scope and try to print the call stack. 

But I agree that there is always this taking it one step deeper of where you make things recoverable.

Reply:
Right, you did reset it, I see now. Sorry, did not look close enough :)